File: app/src/main/java/org/sorz/lab/tinykeepass/DatabaseSetupActivity.java
package org.sorz.lab.tinykeepass;

import android.content.Intent;
import android.content.SharedPreferences;
import android.hardware.fingerprint.FingerprintManager;
import android.net.Uri;
import android.os.Bundle;
import android.text.InputType;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.Spinner;
import android.widget.Toast;

import org.sorz.lab.tinykeepass.keepass.KeePassStorage;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

/**
 *  Provide UI for configure a new db.
 *
 */
public class DatabaseSetupActivity extends BaseActivity {
    final private static String TAG = DatabaseSetupActivity.class.getName();
    final private static int REQUEST_OPEN_FILE = 1;
    public static final int AUTH_METHOD_UNDEFINED = -1;
    public static final int AUTH_METHOD_NONE = 0;
    public static final int AUTH_METHOD_SCREEN_LOCK = 1;
    public static final int AUTH_METHOD_FINGERPRINT = 2;
    public static final String PREF_DB_URL = "db-url";
    public static final String PREF_DB_AUTH_USERNAME = "db-auth-username";
    public static final String PREF_DB_AUTH_REQUIRED = "db-auth-required";
    public static final String PREF_KEY_AUTH_METHOD = "key-auth-method";

    private FingerprintManager fingerprintManager;

    private boolean launchMainActivityAfterSave = false;

    private CheckBox checkBasicAuth;
    private EditText editDatabaseUrl;
    private EditText editAuthUsername;
    private EditText editAuthPassword;
    private EditText editMasterPassword;
    private CheckBox checkShowPassword;
    private Button buttonConfirm;
    private Button buttonOpenFile;
    private Spinner spinnerAuthMethod;
    private ProgressBar progressBar;
    private List<View> disabledViews = new ArrayList<>(8);


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_database_setup);

        fingerprintManager = (FingerprintManager) getSystemService(FINGERPRINT_SERVICE);

        checkBasicAuth = findViewById(R.id.checkBasicAuth);
        editDatabaseUrl = findViewById(R.id.editDatabaseUrl);
        editAuthUsername = findViewById(R.id.editAuthUsername);
        editAuthPassword = findViewById(R.id.editAuthPassword);
        editMasterPassword = findViewById(R.id.editMasterPassword);
        spinnerAuthMethod = findViewById(R.id.spinnerAuthMethod);
        checkShowPassword = findViewById(R.id.checkShowPassword);
        progressBar = findViewById(R.id.progressBar);
        buttonConfirm = findViewById(R.id.buttonConfirm);
        buttonOpenFile = findViewById(R.id.buttonOpenFIle);

        checkBasicAuth.setOnCheckedChangeListener((CompoundButton button, boolean isChecked) -> {
            editAuthUsername.setVisibility(isChecked ? View.VISIBLE : View.GONE);
            editAuthPassword.setVisibility(editAuthUsername.getVisibility());
        });
        checkShowPassword.setOnCheckedChangeListener((CompoundButton button, boolean isChecked) ->
                editMasterPassword.setInputType(isChecked
                        ? InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD
                        : InputType.TYPE_TEXT_VARIATION_PASSWORD | InputType.TYPE_CLASS_TEXT));
        buttonOpenFile.setOnClickListener((View button) -> {
            Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
            intent.addCategory(Intent.CATEGORY_OPENABLE);
            intent.setType("*/*");
            startActivityForResult(intent, REQUEST_OPEN_FILE);
        });
        buttonConfirm.setOnClickListener((View button) -> {
            if (isInputValid())
                submit();
        });

        editDatabaseUrl.setText(getPreferences().getString("db-url", ""));
        editAuthUsername.setText(getPreferences().getString("db-auth-username", ""));
        checkBasicAuth.setChecked(getPreferences().getBoolean(PREF_DB_AUTH_REQUIRED, false));

        // Handle VIEW action
        Intent intent = getIntent();
        if (intent != null && intent.getData() != null) {
            editDatabaseUrl.setText(intent.getData().toString());
            editDatabaseUrl.setEnabled(false);
            buttonOpenFile.setEnabled(false);
            launchMainActivityAfterSave = true;

            //int authMethod = preferences.getInt(PREF_KEY_AUTH_METHOD, AUTH_METHOD_UNDEFINED);
            //if (authMethod != AUTH_METHOD_UNDEFINED)
            // TODO: try open with saved keys.
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        disabledViews.clear();
        checkBasicAuth = null;
        editDatabaseUrl = null;
        editAuthUsername = null;
        editAuthPassword = null;
        editMasterPassword = null;
        spinnerAuthMethod = null;
        checkShowPassword = null;
        buttonConfirm = null;
        buttonOpenFile = null;
    }

    private boolean isInputValid() {
        List<EditText> notEmptyInputs = new ArrayList<>(4);
        notEmptyInputs.add(editDatabaseUrl);
        notEmptyInputs.add(editMasterPassword);
        if (checkBasicAuth.isChecked()) {
            if (!editDatabaseUrl.getText().toString().startsWith("http")) {
                checkBasicAuth.setError(getString(R.string.basic_auth_with_non_http));
                return false;
            }
            notEmptyInputs.add(editAuthUsername);
            notEmptyInputs.add(editAuthPassword);
        }
        for (EditText edit : notEmptyInputs) {
            if (edit.getText().toString().isEmpty()) {
                edit.setError("Cannot be empty");
                return false;
            }
        }

        switch (spinnerAuthMethod.getSelectedItemPosition()) {
            case 0: // no auth
                break;
            case 1: // fingerprint
                if (!fingerprintManager.isHardwareDetected()) {
                    Toast.makeText(this,
                            R.string.no_fingerprint_detected, Toast.LENGTH_LONG).show();
                    return false;
                }
                if (!fingerprintManager.hasEnrolledFingerprints()) {
                    Toast.makeText(this,
                            R.string.no_fingerprint_enrolled, Toast.LENGTH_LONG).show();
                    return false;
                }
                break;
        }

        return true;
    }

    private void submit() {
        Uri uri = Uri.parse(editDatabaseUrl.getText().toString());

        disabledViews.clear();
        disabledViews.add(checkBasicAuth);
        disabledViews.add(editDatabaseUrl);
        disabledViews.add(editAuthUsername);
        disabledViews.add(editAuthPassword);
        disabledViews.add(editMasterPassword);
        disabledViews.add(spinnerAuthMethod);
        disabledViews.add(checkBasicAuth);
        disabledViews.add(buttonConfirm);
        for (View v : disabledViews)
            v.setEnabled(false);
        progressBar.setVisibility(View.VISIBLE);

        String username = null;
        String password = null;
        if (checkBasicAuth.isChecked()) {
            username = editAuthUsername.getText().toString();
            password = editAuthPassword.getText().toString();
        }
        String masterPwd = editMasterPassword.getText().toString();
        new FetchTask(this, uri, masterPwd, username, password).execute();
    }

    private void cancelSubmit() {
        for (View v : disabledViews)
            v.setEnabled(true);
        disabledViews.clear();
        progressBar.setVisibility(View.INVISIBLE);
    }

    static void clearDatabaseConfigs(SharedPreferences preferences) {
        preferences.edit()
                .remove(PREF_DB_URL)
                .remove(PREF_DB_AUTH_USERNAME)
                .remove(PREF_DB_AUTH_REQUIRED)
                .remove(PREF_KEY_AUTH_METHOD)
                .apply();
    }

    private void saveDatabaseConfigs() {
        int authMethod = spinnerAuthMethod.getSelectedItemPosition();
        if (authMethod == AUTH_METHOD_SCREEN_LOCK)
            authMethod = AUTH_METHOD_FINGERPRINT;
        getPreferences().edit()
                .putString(PREF_DB_URL, editDatabaseUrl.getText().toString())
                .putString(PREF_DB_AUTH_USERNAME, editAuthUsername.getText().toString())
                .putBoolean(PREF_DB_AUTH_REQUIRED, checkBasicAuth.isChecked())
                .putInt(PREF_KEY_AUTH_METHOD, authMethod)
                .apply();

        List<String> keys = new ArrayList<>(2);
        keys.add(editMasterPassword.getText().toString());
        keys.add(editAuthPassword.getText().toString());
        saveDatabaseKeys(keys, () -> {
            setResult(RESULT_OK);
            if (launchMainActivityAfterSave)
                startActivity(new Intent(this, MainActivity.class));
            finish();
        }, error -> {
            Toast.makeText(this, error, Toast.LENGTH_SHORT).show();
            getPreferences().edit()
                    .putInt(PREF_KEY_AUTH_METHOD, AUTH_METHOD_UNDEFINED)
                    .apply();
            KeePassStorage.set(this, null);
            cancelSubmit();
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case REQUEST_OPEN_FILE:
                if (resultCode == RESULT_OK && data != null) {
                    Uri uri = data.getData();
                    editDatabaseUrl.setText(uri.toString());
                    checkBasicAuth.setChecked(false);
                }
                break;
            default:
                super.onActivityResult(requestCode, resultCode, data);
                break;
        }
    }


    private static class FetchTask extends FetchDatabaseTask {
        private final WeakReference<DatabaseSetupActivity> activity;

        FetchTask(DatabaseSetupActivity activity, Uri uri, String masterPwd,
                         String username, String password) {
            super(activity, uri, masterPwd, username, password);
            this.activity = new WeakReference<>(activity);
        }

        @Override
        protected void onPostExecute(String error) {
            DatabaseSetupActivity activity = this.activity.get();
            if (activity == null)
                return;

            if (error != null) {
                Toast.makeText(activity, error, Toast.LENGTH_SHORT).show();
                activity.cancelSubmit();
            } else {
                activity.saveDatabaseConfigs();
            }
        }
    }
}

File: app/src/main/java/org/sorz/lab/tinykeepass/DatabaseSetupActivity.kt
package org.sorz.lab.tinykeepass

import android.app.Application
import android.app.KeyguardManager
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.widget.*
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.compose.runtime.getValue
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.platform.setContent
import androidx.core.content.getSystemService
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.lifecycleScope
import androidx.preference.PreferenceManager
import com.kunzisoft.keepass.database.exception.LoadDatabaseException
import org.sorz.lab.tinykeepass.keepass.KeePassStorage
import org.sorz.lab.tinykeepass.ui.Setup
import java.io.IOException


const val AUTH_METHOD_UNDEFINED = -1
const val AUTH_METHOD_NONE = 0
const val AUTH_METHOD_SCREEN_LOCK = 1
const val AUTH_METHOD_FINGERPRINT = 2
const val PREF_DB_URL = "db-url"
const val PREF_DB_AUTH_USERNAME = "db-auth-username"
const val PREF_DB_AUTH_REQUIRED = "db-auth-required"
const val PREF_KEY_AUTH_METHOD = "key-auth-method"

private const val TAG = "DatabaseSetupActivity"

/**
 * Provide UI for configure a new db.
 *
 */
class DatabaseSetupActivity : BaseActivity() {
    private val keyguardManager by lazy { getSystemService() as KeyguardManager? }
    private val viewModel by viewModels<SetupViewModel>()
    private val openDocument = registerForActivityResult(ActivityResultContracts.OpenDocument()) {
        if (it != null) viewModel.path.value = it.toString()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val authSupported = keyguardManager?.isDeviceSecure ?: false
        if (!authSupported) viewModel.enableAuth.value = false

        // Handle VIEW action
        val viewActionUri = intent?.data?.also {
            viewModel.path.value = it.toString()
            viewModel.basicAuth.value = BasicAuthCfg(enabled = false)
        }
        val forViewAction = viewActionUri != null

        // Setup UI
        setContent {
            val path by viewModel.path.observeAsState("")
            val basicAuth by viewModel.basicAuth.observeAsState(BasicAuthCfg())
            val enableAuth by viewModel.enableAuth.observeAsState(false)
            val masterPassword by viewModel.masterPassword.observeAsState("")
            val state by viewModel.state.observeAsState(SetupState.EDITING)

            Setup(
                path = path,
                onPathChange = if (forViewAction) null else viewModel.path::setValue,
                onOpenFile = if (forViewAction) null else {{ openDocument.launch(arrayOf("*/*")) }},
                basicAuthCfg = basicAuth,
                onBasicAuthCfgChange = viewModel.basicAuth::setValue,
                enableAuth = enableAuth,
                onEnabledAuthChange = if (authSupported) viewModel.enableAuth::setValue else null,
                masterPassword = masterPassword,
                onMasterPasswordChange = viewModel.masterPassword::setValue,
                onSubmit = {
                    viewModel.state.value = SetupState.VALIDATING
                    doFetchDatabase()
                },
                isInProgress = state != SetupState.EDITING,
            )
        }

        // Saving config after database fetched
        viewModel.state.observe(this) { state ->
            if (state != SetupState.DONG_WITHOUT_ERROR) return@observe
            val authMethod = if (viewModel.enableAuth.value!!) AUTH_METHOD_FINGERPRINT else 0
            val basicAuth = viewModel.basicAuth.value!!
            preferences.edit()
                    .putString(PREF_DB_URL, viewModel.path.value!!)
                    .putString(PREF_DB_AUTH_USERNAME, basicAuth.username)
                    .putBoolean(PREF_DB_AUTH_REQUIRED, basicAuth.enabled)
                    .putInt(PREF_KEY_AUTH_METHOD, authMethod)
                    .apply()
            val keys = arrayListOf(viewModel.masterPassword.value!!, basicAuth.password)
            lifecycleScope.launchWhenStarted {
                doSaveDatabaseKeys(keys, forViewAction)
            }
        }
    }

    private suspend fun doSaveDatabaseKeys(keys: List<String>, launchMainActivity: Boolean) {
        try {
            saveDatabaseKeys(keys)
        } catch (err: AuthKeyError) {
            Toast.makeText(this, err.message!!, Toast.LENGTH_SHORT).show()
            preferences.edit()
                    .putInt(PREF_KEY_AUTH_METHOD, AUTH_METHOD_UNDEFINED)
                    .apply()
            KeePassStorage.set(applicationContext, null)
            viewModel.state.value = SetupState.EDITING
            return
        }
        setResult(RESULT_OK)
        if (launchMainActivity)
            startActivity(Intent(applicationContext, MainActivity::class.java))
        finish()
    }

    private fun doFetchDatabase() = lifecycleScope.launchWhenCreated {
        val path = Uri.parse(viewModel.path.value!!)
        val masterPwd = viewModel.masterPassword.value!!
        val basicAuth = viewModel.basicAuth.value!!
        try {
            fetchDatabase(baseContext, path, masterPwd, basicAuth)
        } catch (err: IOException) {
            Log.w(TAG, "error on fetch db", err)
            val msg = getString(R.string.fail_to_sync, err.localizedMessage ?: "I/O error")
            Toast.makeText(baseContext, msg, Toast.LENGTH_SHORT).show()
            viewModel.state.value = SetupState.EDITING
            return@launchWhenCreated
        }  catch (err: LoadDatabaseException) {
            Log.w(TAG, "error on open db", err)
            val msg = getString(R.string.open_db_dialog_fail)
            Toast.makeText(baseContext, msg, Toast.LENGTH_SHORT).show()
            viewModel.state.value = SetupState.EDITING
            return@launchWhenCreated
        }
        viewModel.state.value = SetupState.DONG_WITHOUT_ERROR
    }

}

enum class SetupState {
    EDITING,
    VALIDATING,
    DONG_WITHOUT_ERROR,
}

class SetupViewModel(val app: Application) : AndroidViewModel(app) {
    private val prefs = PreferenceManager.getDefaultSharedPreferences(app)
    val path = MutableLiveData(prefs.getString(PREF_DB_URL, "") ?: "")
    val basicAuth = MutableLiveData(BasicAuthCfg(
        enabled = prefs.getBoolean(PREF_DB_AUTH_REQUIRED, false),
        username = prefs.getString(PREF_DB_AUTH_USERNAME, "") ?: "",
    ))
    val masterPassword = MutableLiveData("")
    val enableAuth = MutableLiveData(prefs.getBoolean(PREF_DB_AUTH_REQUIRED, false))
    val state = MutableLiveData(SetupState.EDITING)
}

fun clearDatabaseConfigs(preferences: SharedPreferences) {
    preferences.edit()
            .remove(PREF_DB_URL)
            .remove(PREF_DB_AUTH_USERNAME)
            .remove(PREF_DB_AUTH_REQUIRED)
            .remove(PREF_KEY_AUTH_METHOD)
            .apply()
}